app/v2/controllers/ErrorTranslator.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2.controllers
+
+import cats.data.EitherT
+import v2.models.errors.BaseError
+import v2.models.errors.FailedToValidateError
+import v2.models.errors.FailedToValidateError.InvalidMessageTypeError
+import v2.models.errors.FailedToValidateError.OtherError
+import v2.models.errors.FailedToValidateError.SchemaFailedToValidateError
+import v2.models.errors.PersistenceError
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+trait ErrorTranslator {
+
+  implicit class ErrorConverter[E, A](value: EitherT[Future, E, A]) {
+
+    def convertError(implicit c: Converter[E], ec: ExecutionContext): EitherT[Future, BaseError, A] =
+      value.leftMap(c.convert)
+  }
+
+  trait Converter[E] {
+    def convert(input: E): BaseError
+  }
```

Comment:
I like this, but oddly enough *THIS* is where I think we're starting to require an enforcement of a base error type that this can all depend on. Base Error has always been more of a presentation-layer approach to errors which is fine, but this would be stricter if it enforced a conversion from EitherT[Future, NewBaseError, A] with convertError being EitherT[Future, NewBaseError, A] and convert can then be def convert(input: NewBaseError): PresentationError - which (I think) at least has more accurate nomenclature